pkg/ring/model_test.go

```diff
@@ -240,3 +240,78 @@ func TestDesc_getTokensByZone(t *testing.T) {
 		})
 	}
 }
+
+func TestDesc_RingsCompare(t *testing.T) {
+	tests := map[string]struct {
+		r1, r2   *Desc
+		expected CompareResult
+	}{
+		"nil rings": {
+			r1:       nil,
+			r2:       nil,
+			expected: Equal,
+		},
+		"one nil, one empty ring": {
+			r1:       nil,
+			r2:       &Desc{Ingesters: map[string]IngesterDesc{}},
+			expected: Equal,
+		},
+		"two empty rings": {
+			r1:       &Desc{Ingesters: map[string]IngesterDesc{}},
+			r2:       &Desc{Ingesters: map[string]IngesterDesc{}},
+			expected: Equal,
+		},
+		"same single instance": {
+			r1:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1"}}},
+			r2:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1"}}},
+			expected: Equal,
+		},
+		"same single instance, different timestamp": {
+			r1:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1", Timestamp: 123456}}},
+			r2:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1", Timestamp: 789012}}},
+			expected: EqualInstancesAndTokens,
+		},
+		"same single instance, different state": {
+			r1:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1", State: ACTIVE}}},
+			r2:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1", State: JOINING}}},
+			expected: EqualInstancesAndTokens,
+		},
+		"instance in different zone": {
+			r1:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1", Zone: "one"}}},
+			r2:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1", Zone: "two"}}},
+			expected: Different,
+		},
+		"same instance, different address": {
+			r1:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1"}}},
+			r2:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr2"}}},
+			expected: Different,
+		},
+		"more instances in one ring": {
+			r1:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1"}, "ing2": {Addr: "ing2"}}},
+			r2:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1"}}},
+			expected: Different,
+		},
+		"different tokens": {
+			r1:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1", Tokens: []uint32{1, 2, 3}}}},
+			r2:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1"}}},
+			expected: Different,
+		},
+		"different tokens 2": {
+			r1:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1", Tokens: []uint32{1, 2, 3}}}},
+			r2:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1", Tokens: []uint32{1, 2, 4}}}},
+			expected: Different,
+		},
+		"same number of instances, using different IDs": {
+			r1:       &Desc{Ingesters: map[string]IngesterDesc{"ing1": {Addr: "addr1", Tokens: []uint32{1, 2, 3}}}},
+			r2:       &Desc{Ingesters: map[string]IngesterDesc{"ing2": {Addr: "addr1", Tokens: []uint32{1, 2, 4}}}},
```

Comment:
This test should have same exact tokens.

```suggestion
			r2:       &Desc{Ingesters: map[string]IngesterDesc{"ing2": {Addr: "addr1", Tokens: []uint32{1, 2, 3}}}},
```

Reply:
Ooh, good point!